crates/trie/sparse/src/trie.rs

```diff
@@ -451,59 +416,48 @@ impl RevealedSparseTrie {
         masks: TrieMasks,
         retain_updates: bool,
     ) -> SparseTrieResult<Self> {
-        let mut this = Self {
-            nodes: HashMap::default(),
-            branch_node_tree_masks: HashMap::default(),
-            branch_node_hash_masks: HashMap::default(),
-            values: HashMap::default(),
-            prefix_set: PrefixSetMut::default(),
-            rlp_buf: Vec::new(),
-            updates: None,
-        }
-        .with_updates(retain_updates);
-        this.reveal_node(Nibbles::default(), root, masks)?;
-        Ok(this)
-    }
-}
-
-impl RevealedSparseTrie {
-    /// Sets the fields of this `RevealedSparseTrie` to the fields of the input
-    /// `SparseTrieState`.
-    ///
-    /// This is meant for reusing the allocated maps contained in the `SparseTrieState`.
-    ///
-    /// Copies over any existing nodes, branch masks, and values.
-    pub fn use_allocated_state(&mut self, mut other: SparseTrieState) {
-        for (path, node) in self.nodes.drain() {
-            other.nodes.insert(path, node);
-        }
-        for (path, mask) in self.branch_node_tree_masks.drain() {
-            other.branch_node_tree_masks.insert(path, mask);
-        }
-        for (path, mask) in self.branch_node_hash_masks.drain() {
-            other.branch_node_hash_masks.insert(path, mask);
-        }
-        for (path, value) in self.values.drain() {
-            other.values.insert(path, value);
-        }
-
-        self.nodes = other.nodes;
-        self.branch_node_tree_masks = other.branch_node_tree_masks;
-        self.branch_node_hash_masks = other.branch_node_hash_masks;
-        self.values = other.values;
+        Self::default().with_root(root, masks, retain_updates)
     }
 
     /// Configures the trie to retain information about updates.
     ///
     /// If `retain_updates` is true, the trie will record branch node updates and deletions.
     /// This information can then be used to efficiently update an external database.
     pub fn with_updates(mut self, retain_updates: bool) -> Self {
-        if retain_updates {
-            self.updates = Some(SparseTrieUpdates::default());
-        }
+        self.updates = retain_updates.then(|| {
```

Comment:
hmm, I think this would change the existing behaviour. If you re-use a sparse trie that had `with_updates(true)`, but now you don't call `with_updates(true)`, it will still collect updates, as the pre-allocated updates vec is `Some`.

Reply:
I think you're correct... it wasn't a problem previously because we didn't directly re-use the RevealedSparseTrie, so updates would always start as None. I think the solution will be to have reset/clear always set updates to None for now, and revert this change. If we want to re-use the updates buffer we can figure out something for that later, we aren't doing it up till now anyway.

Reply:
sgtm